### PR TITLE
Rocprim integration

### PR DESCRIPTION
**Pull request by iotamudelta**
After talking to @dagamayank , here is a pull request of the current status (w/o the kernels like reduction blocked by missing rocPRIM API).

We probably want to wait with merging for CI - I'd just like to get a review on code. Thanks!

### CODE REVIEW

File: tensorflow/core/util/rocm_kernel_helper.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#if TENSORFLOW_USE_ROCM
```
Comment (whchung) [inline tensorflow/core/util/rocm_kernel_helper.h]: Please kill this file. There is another gpu_kernel_helper which does the same thing

File: tensorflow/core/kernels/histogram_op_gpu.cu.cc [hunk 2]
```diff
@@ -30,10 +30,9 @@ limitations under the License.
 #include "tensorflow/core/kernels/histogram_op.h"
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/util/gpu_kernel_helper.h"
 #if GOOGLE_CUDA
 #include "tensorflow/core/util/cuda_kernel_helper.h"
```
Comment (whchung) [inline tensorflow/core/kernels/histogram_op_gpu.cu.cc]: This is incorrect . In GOOGLE_CUDA path there is no cuda_kernel_helper.h

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 3]
```diff
@@ -45,6 +45,23 @@ limitations under the License.
 #define GPU_AXIS_KERNEL_LOOP(i, n, axis) \
   for (int i : ::tensorflow::GpuGridRange##axis<int>(n))
 
+#if TENSORFLOW_USE_ROCM
+
+#define cub hipcub
```
Comment (whchung) [inline tensorflow/core/util/gpu_kernel_helper.h]: I’ll need to sit behind a computer to check this header again. Most of cuda to hipnchanges don’t sound correct to me

File: tensorflow/core/kernels/histogram_op_gpu.cu.cc (hunk 2 above)
Comment (iotamudelta) [reply tensorflow/core/kernels/histogram_op_gpu.cu.cc]: Please note that the include for cuda_kernel_helper.h existed before my changes, I simply put it behind the GOOGLE_CUDA macro. So I am not sure why it would be incorrect?

File: tensorflow/core/util/gpu_kernel_helper.h (hunk 3 above)
Comment (iotamudelta) [reply tensorflow/core/util/gpu_kernel_helper.h]: Please elaborate. I am happy to add more advanced macros, however, these simple define changes do compile correctly.

File: tensorflow/core/kernels/histogram_op_gpu.cu.cc (hunk 2 above)
Comment (whchung) [reply tensorflow/core/kernels/histogram_op_gpu.cu.cc]: cuda_kernel_helper.h is not in this repository anymore

Comment (iotamudelta) [reply tensorflow/core/kernels/histogram_op_gpu.cu.cc]: I see, did you move it to gpu_kernel_helper.h when porting? Anyways, will replace the header inclusion here and elsewhere then with gpu_kernel_helper.

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 4]
```diff
@@ -45,6 +45,23 @@ limitations under the License.
...
+#define cudaSuccess hipSuccess
+#define cudaGetErrorString hipGetErrorString
+#define cudaStream_t hipStream_t
+#define cudaGetLastError hipGetLastError
+#define cudaError hipError
+
+#define CUDA_1D_KERNEL_LOOP(i, n)                                   \
```
Comment (whchung) [inline tensorflow/core/util/gpu_kernel_helper.h]: Please help check if this macro can be abolished and use `GPU_1D_KERNEL_LOOP` also defined in this header file.

File: tensorflow/core/util/gpu_kernel_helper.h [hunk 5]
```diff
@@ -45,6 +45,23 @@ limitations under the License.
...
+    for (size_t i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+        i += blockDim.x * gridDim.x)
+
+#define GetCudaStream(context) context->eigen_gpu_device().stream()
```
Comment (whchung) [inline tensorflow/core/util/gpu_kernel_helper.h]: `GetGpuStream`

File: tensorflow/core/util/gpu_kernel_helper.h (hunk 3 above)
Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: One of the goal is to try restrict exposure to `cuda` for less potential legal issue. I'd feel more comfortable putting `#if/#elif` blocks in the operators.

**Comment by whchung** [comment]
@iotamudelta Please also help revise https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream/rocm_docs/core_kernels.md to help us track where we are for those ops dependent on `cub`/`rocPRIM`

File: tensorflow/core/util/gpu_kernel_helper.h (hunk 5 above)
Comment (iotamudelta) [reply tensorflow/core/util/gpu_kernel_helper.h]: Did you also remove the GetCudaStream macro when you removed cuda_kernel_helper.h ?

Comment (whchung) [reply tensorflow/core/util/gpu_kernel_helper.h]: I believe so. You might want to double check though.

**Comment by iotamudelta** [comment]
@whchung could you check if this is now OK? Thanks!

**Comment by whchung** [comment]
Since reduction/L2 loss/softmax are not done yet could you help leave pointers to relevant discussions on rocPRIM repo here?

**Comment by iotamudelta** [comment]
There is my rocprim-master branch which includes (based on the old header defines, will update) my changes to get reduction/l2loss/softmax compiled. It depends on the rocPRIM snapshot they provided but I'll update next week when they release a new milestone. I have been told that there at least three bugs in the snapshot which will be fixed in the milestone, so below hopefully will improve as well.

The remaining issues are:
* reduction unit tests don't pass - but they also don't pass for me before the rocPRIM changes so we need to look into this once we have a "stable" rocPRIM
* complex data types don't work for the reduction, rocPRIM will comment on that next week
* topk originally had issues with radix_sort for Eigen::half, there is a workaround but then there are other issues appearing, no clue yet what is causing that/how to fix
* where: there is a sizeof on an incomplete type inside rocPRIM- rocPRIM has been informed
* dynamic_partition: gives me a no matching function for call to object error, it relies on the reduction kernel so this is a bit surprising - rocPRIM has been informed.

**Comment by iotamudelta** [comment]
See [rocprim-master branch](https://github.com/iotamudelta/tensorflow-upstream/tree/rocprim-master), now with the latest macro/type alias definitions.
